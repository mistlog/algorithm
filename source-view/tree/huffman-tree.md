```typescript
export class HuffmanTree {
    m_Root: HuffmanNode;
}
```

```typescript
export class HuffmanNode extends NumberNode {}
```

```typescript
<HuffmanTree /> +
    function constructor(this: HuffmanTree, values: Array<number>) {
        <PrepareQueue />;
        <CombineNodes />;
        <SetRoot />;
    };
```

```typescript
function PrepareQueue(values: Array<number>) {
    const queue = new PriorityQueue<HuffmanNode>({ comparator: (a: HuffmanNode, b: HuffmanNode) => a.m_Info - b.m_Info, initialValues: values.map(each => new NumberNode(each)) });
}
```

```typescript
function SetRoot(this: HuffmanTree, queue: PriorityQueue<HuffmanNode>) {
    this.m_Root = queue.dequeue();
}
```

```typescript
function CombineNodes(queue: PriorityQueue<HuffmanNode>) {
    while (queue.length !== 1) {
        const left = queue.dequeue();
        const right = queue.dequeue();
        const node = new HuffmanNode(left.m_Info + right.m_Info, left, right);
        queue.queue(node);
    }
}
```