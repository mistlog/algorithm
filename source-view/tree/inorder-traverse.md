```typescript
export class InorderTraverse {
    m_Stack: Array<StringNode>;
    m_Visited: Array<string>;
}
```

```typescript
<InorderTraverse /> +
    function Traverse(this: InorderTraverse & IInorderTraverse, root: StringNode) {
        let current = root;
        for (;;) {
            if (current) {
                <AddCurrentToTobeVisited />;
            } else {
                <CheckIfFinished />;
                <SetCurrentAsTobeVisited />;
                <VisitStringNode />;
            }
        }
    };
```

```typescript
function AddCurrentToTobeVisited(this: InorderTraverse, current: StringNode) {
    this.m_Stack.push(current);
    current = current.m_Left as StringNode;
}
```

```typescript
function CheckIfFinished(this: InorderTraverse) {
    if (this.m_Stack.length === 0) {
        return;
    }
}
```

```typescript
function VisitStringNode(this: IInorderTraverse, current: StringNode) {
    this.Visit(current);
    current = current.m_Right as StringNode;
}
```

```typescript
function SetCurrentAsTobeVisited(this: InorderTraverse, current: StringNode) {
    current = this.m_Stack.pop();
}
```

```typescript
<InorderTraverse /> +
    function constructor(this: InorderTraverse) {
        this.m_Stack = new Array();
        this.m_Visited = new Array();
    };
```

```typescript
export interface IInorderTraverse {
    Visit(node: StringNode): void;
}
```

```typescript
<InorderTraverse /> +
    function Visit(this: InorderTraverse, node: StringNode) {
        this.m_Visited.push(node.m_Info);
    };
```

# Reference

This algo is adapted from TAOCP I 2.3.1