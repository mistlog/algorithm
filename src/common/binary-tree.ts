export class Node
{
    m_Left: Node;
    m_Right: Node;

    constructor(left?: Node, right?: Node)
    {
        this.m_Left = left || null;
        this.m_Right = right || null;
    }
}

export class StringNode extends Node
{
    m_Info: string;

    constructor(info: string, left?: Node, right?: Node)
    {
        super(left, right);
        this.m_Info = info;
    }
}

export class NumberNode extends Node
{
    m_Info: number;

    constructor(info: number, left?: Node, right?: Node)
    {
        super(left, right);
        this.m_Info = info;
    }
}