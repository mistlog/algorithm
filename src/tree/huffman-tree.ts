import { NumberNode } from "../common/binary-tree";
import { default as PriorityQueue } from "ts-priority-queue";
export class HuffmanTree {
  m_Root: HuffmanNode;

  constructor(values: Array<number>) {
    const queue = new PriorityQueue<HuffmanNode>({
      comparator: (a: HuffmanNode, b: HuffmanNode) => a.m_Info - b.m_Info,
      initialValues: values.map(each => new NumberNode(each))
    });

    while (queue.length !== 1) {
      const left = queue.dequeue();
      const right = queue.dequeue();
      const node = new HuffmanNode(left.m_Info + right.m_Info, left, right);
      queue.queue(node);
    }

    this.m_Root = queue.dequeue();
  }

}
export class HuffmanNode extends NumberNode {}