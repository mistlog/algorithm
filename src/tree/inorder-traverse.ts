import { StringNode } from "../common/binary-tree";
export class InorderTraverse {
  m_Stack: Array<StringNode>;
  m_Visited: Array<string>;

  Traverse(root: StringNode) {
    let current = root;

    for (;;) {
      if (current) {
        this.m_Stack.push(current);
        current = (current.m_Left as StringNode);
      } else {
        if (this.m_Stack.length === 0) {
          return;
        }

        current = this.m_Stack.pop();
        this.Visit(current);
        current = (current.m_Right as StringNode);
      }
    }
  }

  constructor() {
    this.m_Stack = new Array<StringNode>();
    this.m_Visited = new Array<string>();
  }

  Visit(node: StringNode) {
    this.m_Visited.push(node.m_Info);
  }

}
export interface IInorderTraverse {
  Visit(node: StringNode): void;
}
;