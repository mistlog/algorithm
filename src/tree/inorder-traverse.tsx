import { StringNode } from "../common/binary-tree";

export class InorderTraverse
{
    m_Stack: Array<StringNode>;
    m_Visited: Array<string>;
}

<InorderTraverse /> + function Traverse(this: InorderTraverse & IInorderTraverse, root: StringNode)
{
    let current = root;

    for (; ;)
    {
        if (current)
        {
            //@ts-ignore
            <AddCurrentToTobeVisited />;
        }
        else
        {
            <CheckIfFinished />;

            //@ts-ignore
            <SetCurrentAsTobeVisited />;

            //@ts-ignore
            <VisitStringNode />;
        }
    }
};

function AddCurrentToTobeVisited(this: InorderTraverse, current: StringNode)
{
    this.m_Stack.push(current)
    current = current.m_Left as StringNode;
}

function CheckIfFinished(this: InorderTraverse)
{
    if (this.m_Stack.length === 0)
    {
        return;
    }
}

function VisitStringNode(this: IInorderTraverse, current: StringNode)
{
    this.Visit(current);
    current = current.m_Right as StringNode;
}

function SetCurrentAsTobeVisited(this: InorderTraverse, current: StringNode)
{
    current = this.m_Stack.pop();
}

<InorderTraverse /> + function constructor(this: InorderTraverse)
{
    this.m_Stack = new Array<StringNode>();
    this.m_Visited = new Array<string>();
};

export interface IInorderTraverse
{
    Visit(node: StringNode): void;
};

<InorderTraverse /> + function Visit(this: InorderTraverse, node: StringNode)
{
    this.m_Visited.push(node.m_Info);
}

/*
# Reference

This algo is adapted from TAOCP I 2.3.1

*/