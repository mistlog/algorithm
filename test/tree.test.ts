import { StringNode } from "../src/common/binary-tree";
import { InorderTraverse } from "../src/tree/inorder-traverse";
import { HuffmanTree } from "../src/tree/huffman-tree";

describe("binary tree", () =>
{
    /**
     * example taken from TAOCP I 2.3.1
     */
    test("traverse.inorder", () =>
    {
        /**
         * 
         */
        const A = new StringNode("A");
        const B = A.m_Left = new StringNode("B");
        const D = B.m_Left = new StringNode("D");
        const C = A.m_Right = new StringNode("C");
        const E = C.m_Left = new StringNode("E");
        const G = E.m_Right = new StringNode("G");
        const F = C.m_Right = new StringNode("F");
        const H = F.m_Left = new StringNode("H");
        const J = F.m_Right = new StringNode("J");

        /**
         * 
         */
        const algo = new InorderTraverse();
        algo.Traverse(A);
        expect(algo.m_Visited).toEqual(["D", "B", "A", "E", "G", "C", "H", "F", "J"]);
    })

    /**
     * example taken from TAOCP I 2.3.4.5
     */
    test("huffman", () =>
    {
        const tree = new HuffmanTree([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]);
        expect(tree.m_Root).toMatchSnapshot();
    })
})