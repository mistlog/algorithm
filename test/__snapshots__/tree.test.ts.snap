// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binary tree huffman 1`] = `
HuffmanNode {
  "m_Info": 238,
  "m_Left": HuffmanNode {
    "m_Info": 95,
    "m_Left": HuffmanNode {
      "m_Info": 42,
      "m_Left": NumberNode {
        "m_Info": 19,
        "m_Left": null,
        "m_Right": null,
      },
      "m_Right": NumberNode {
        "m_Info": 23,
        "m_Left": null,
        "m_Right": null,
      },
    },
    "m_Right": HuffmanNode {
      "m_Info": 53,
      "m_Left": HuffmanNode {
        "m_Info": 24,
        "m_Left": NumberNode {
          "m_Info": 11,
          "m_Left": null,
          "m_Right": null,
        },
        "m_Right": NumberNode {
          "m_Info": 13,
          "m_Left": null,
          "m_Right": null,
        },
      },
      "m_Right": NumberNode {
        "m_Info": 29,
        "m_Left": null,
        "m_Right": null,
      },
    },
  },
  "m_Right": HuffmanNode {
    "m_Info": 143,
    "m_Left": HuffmanNode {
      "m_Info": 65,
      "m_Left": NumberNode {
        "m_Info": 31,
        "m_Left": null,
        "m_Right": null,
      },
      "m_Right": HuffmanNode {
        "m_Info": 34,
        "m_Left": HuffmanNode {
          "m_Info": 17,
          "m_Left": NumberNode {
            "m_Info": 7,
            "m_Left": null,
            "m_Right": null,
          },
          "m_Right": HuffmanNode {
            "m_Info": 10,
            "m_Left": NumberNode {
              "m_Info": 5,
              "m_Left": null,
              "m_Right": null,
            },
            "m_Right": HuffmanNode {
              "m_Info": 5,
              "m_Left": NumberNode {
                "m_Info": 2,
                "m_Left": null,
                "m_Right": null,
              },
              "m_Right": NumberNode {
                "m_Info": 3,
                "m_Left": null,
                "m_Right": null,
              },
            },
          },
        },
        "m_Right": NumberNode {
          "m_Info": 17,
          "m_Left": null,
          "m_Right": null,
        },
      },
    },
    "m_Right": HuffmanNode {
      "m_Info": 78,
      "m_Left": NumberNode {
        "m_Info": 37,
        "m_Left": null,
        "m_Right": null,
      },
      "m_Right": NumberNode {
        "m_Info": 41,
        "m_Left": null,
        "m_Right": null,
      },
    },
  },
}
`;
